src/middlewared/middlewared/plugins/smart.py

```diff
@@ -1,8 +1,154 @@
+                f'{schema}.disks',
+                f'The following disks are invalid: {", ".join(invalid_disks)}'
+            )
+
+        return verrors
+
+    @accepts(
+        Dict(
+            'smart_task_create',
+            Str('desc'),
+            List('disks', items=[Str('disk')], required=True),
+            Str('dayweek'),
+            Str('month'),
+            Str('daymonth'),
+            Str('hour'),
+            Str('type', enum=['L', 'S', 'C', 'O'], required=True),
+            register=True
+        )
+    )
+    async def do_create(self, data):
+        verrors = await self.validate_data(data, 'smart_test_create')
+
+        if not data.get('disks'):
+            verrors.add(
+                'smart_test_create.disks',
+                'This field is required'
+            )
+
+        if verrors:
+            raise verrors
+
+        data['id'] = await self.middleware.call(
+            'datastore.insert',
+            self._config.datastore,
+            data,
+            {'prefix': self._config.datastore_prefix}
+        )
+
+        await self.middleware.call(
+            'service.restart',
+            'smartd',
+            {'onetime': False}
+        )
+
+        return data
+
+    @accepts(
+        Int('id', validators=[Range(min=1)]),
+        Patch('smart_task_create', 'smart_task_update', ('attr', {'update': True}))
+    )
+    async def do_update(self, id, data):
+        old = await self.query(filters=[('id', '=', id)], options={'get': True})
+        new = old.copy()
+        new.update(data)
+
+        new_disks = [disk for disk in new['disks'] if disk not in old['disks']]
+        deleted_disks = [disk for disk in old['disks'] if disk not in new['disks']]
+        if old['type'] == new['type']:
+            new['disks'] = new_disks
+        verrors = await self.validate_data(new, 'smart_test_update')
+
+        new['disks'] = [disk for disk in chain(new_disks, old['disks']) if disk not in deleted_disks]
+
+        if not new.get('disks'):
+            verrors.add(
+                'smart_test_update.disks',
+                'This field is required'
+            )
+
+        if verrors:
+            raise verrors
+
+        await self.middleware.call(
+            'datastore.update',
+            self._config.datastore,
+            id,
+            new,
+            {'prefix': self._config.datastore_prefix}
+        )
+
+        await self.middleware.call(
+            'service.restart',
+            'smartd',
+            {'onetime': False}
+        )
+
+        return await self.query(filters=[('id', '=', id)], options={'get': True})
+
+    @accepts(
+        Int('id')
+    )
+    async def do_delete(self, id):
+        response = await self.middleware.call(
+            'datastore.delete',
+            self._config.datastore,
+            id
+        )
+
+        await self.middleware.call(
+            'service.restart',
+            'smartd',
+            {'onetime': False}
+        )
```

Comment:
Service is already restarted on tasks/models.py:delete

Reply:
@william-gr isn't this something we should try moving out of Django while doing these migrations?

Reply:
@themylogin I thought of that but then we would need an override of the delete() method on ModelForm. I think we should, but in a new PR, carefully, because so far we dont have anything use middleware delete (except boot env I think) and that might unravel a new set of bugs.